Webpack 2 support

renarsvilnis opened the pull request:
Initial pull request for #38 

Stuff that has been done:
- Add 'src/helpers.js' for commonly used functions
- Add JSDOC documentation for various functions
- Add support for Webpack v2
- Add enhanced support for Webpack v1 by reading 'resolve.root', 'resolve.fallback'

> Pull request needs feedback and is unfinished
> ⚠️  **Don't merge, yet!**

I'm worried that the enhanced Webpack v1 and Webpack v2 support might not work as they hold absolute paths (e.g. `path.join(__dirname, 'app')`) which in the `src/helpers.js@lookupLocal` function might not resolve correctly.

I'd love to see some help on adding documentation on remaining functions and especially `lookupLocal` would help me understand the library and finish the pull request.

nkt commented on src/helpers.js:
```diff
@@ -0,0 +1,43 @@
+'use babel';
```
Please create `utils/` folder and place every helper separately.

nkt commented on src/webpack-lookup.js:
```diff
@@ -0,0 +1,97 @@
+'use babel';
```
Move this to `utils/`

renarsvilnis commented:
### Update

Did a overall file structure refactor:
- renamed `src` folder to `lib` (*common npm practice*)
- added top-level `index.js` that just exports the plugin instance (*common npm practice*)
- moved helper/private functions from `completion-provider.js` to `utils`.

I can also try adding testing and continuous-integration using `mocha`,  `chai`, `istanbul`, `travis`.

Should I close this branch and make a pull request to `v2` branch?

nkt commented:
Atom has own test infrastructure. [Atom Linter](https://github.com/steelbrain/linter) as example of package with tests.
I prefer CircleCI, but travis is ok.

 > Should I close this branch and make a pull request to v2 branch?

No, I'll update master.

nkt commented:
@renarsvilnis please add `circle.yml`:
```yaml
machine:
  node:
    version: 6
```

renarsvilnis commented:
Added `circle.yml`.

nkt commented:
@renarsvilnis is it still WIP?

renarsvilnis commented:
Yes, haven't had time to figure out how `lookupLocal` function works and if it would do mad-stuff when passed absolute paths from the webpack config.

nkt commented:
@renarsvilnis I guess `lookupLocal`, actually works not fine right now.
The algorithm should be:
 - split directory and filename ('/foo' and 'ba'), ('' and 'fo')
 - concat current file directory and split directory from prev step
 - lookup names in dirname from prev step
 - filter names using filename prefix from first step

But first step works not fine. For example, if prefix is `'..`, `'./'`, `'../..'` it's becomes `''`. So if you have better implementation, I'll glad to see it.

nkt commented:
@renarsvilnis are you still interested in this changes?

renarsvilnis commented:
@nkt Yeah, but I quite busy right now, can get back to it in a month.

MatthieuLemoine commented:
@renarsvilnis are you still willing to work on this or should I give it a try ?

renarsvilnis commented:
@MatthieuLemoine no, have forgotten about it. You can continue on it.

jonyeezs commented:
Can this be close, @nkt @renarsvilnis ? @MatthieuLemoine seems to have solved it on PR #64 ?

jonyeezs commented:
Superseded by #64
